### PR TITLE
Added: Integration between Extraction AIs and Konfuzio Documents

### PR DESCRIPTION
**Pull request by da-konfuzio**

### CODE REVIEW

File: konfuzio_sdk/trainer/information_extraction.py [hunk 1]
```diff
@@ -1718,6 +1719,86 @@ def generate_feature_dict_from_occurence_dict(occurence_dict, catchphrase_list,
     return _dict
 
 
+def add_extractions_as_annotations(
```
Comment (atraining) [inline konfuzio_sdk/trainer/information_extraction.py:1722]: perhaps we can add some basic tests

File: konfuzio_sdk/trainer/information_extraction.py [hunk 2]
```diff
@@ -1718,6 +1719,86 @@ def generate_feature_dict_from_occurence_dict(occurence_dict, catchphrase_list,
...
+                document=document,
+                label=label,
+                confidence=annotation['Accuracy'],
+                label_set=label_set,
+                annotation_set=annotation_set,
+                spans=[Span(start_offset=annotation['Start'], end_offset=annotation['End'])]
+            )
+
+
+def extraction_result_to_document(document: Document, extraction_result: dict) -> Document:
```
Comment (atraining) [inline konfuzio_sdk/trainer/information_extraction.py:1747]: perhaps we can add some basic tests

Comment (da-konfuzio) [inline konfuzio_sdk/trainer/information_extraction.py:1747]: todo: add tests for this

File: tests/trainer/test_information_extraction.py [hunk 3]
```diff
@@ -270,6 +272,75 @@ def test_7_upload_ai_model(self):
...
+        self.pipeline.test_documents = self.pipeline.category.test_documents()[:1]
+
+    def test_2_find_regex(self):
+        """Fit the tokenizer."""
+        for label in self.pipeline.category.labels:
+            for regex in label.find_regex(category=self.pipeline.category):
+                self.pipeline.tokenizer.tokenizers.append(RegexTokenizer(regex=regex))
+        eval_list = []
+        for document in self.pipeline.test_documents:
+            tokenized_doc = self.pipeline.tokenizer.tokenize(document)
```
Comment (da-konfuzio) [inline tests/trainer/test_information_extraction.py:298]: todo deepcopy

File: tests/trainer/test_information_extraction.py [hunk 4]
```diff
@@ -270,6 +272,82 @@ def test_7_upload_ai_model(self):
...
+    def test3_perfect_tokenizer_coverage(self):
+        """Check 100% tokenizer coverage."""
+        from copy import deepcopy
+
+        # todo integrate into tokenizer.evaluate(documents=pipeline.test_documents) -> Evaluation
+        eval_list = []
+        for document in self.pipeline.test_documents:
+            tokenized_doc = self.pipeline.tokenizer.tokenize(deepcopy(document))
+            eval_list.append((document, tokenized_doc))
+        tokenizer_eval = Evaluation(eval_list)
```
Comment (da-konfuzio) [inline tests/trainer/test_information_extraction.py:307]: @atraining before I integrate this into the tokenizer.evaluate method, which would change many related testcases, any reason I shouldn't do it?

Comment (atraining) [reply tests/trainer/test_information_extraction.py:307]: There is no tokenizer.evaluate().

File: tests/trainer/test_information_extraction.py [hunk 5]
```diff
@@ -67,63 +74,63 @@ def display_top(snapshot, key_type='lineno', limit=30):
...
+        #     doc.set_offline()
 
     def test_2_make_features(self):
         """Make sure the Data and Pipeline is configured."""
         self.pipeline.df_train, self.pipeline.label_feature_list = self.pipeline.feature_function(
             documents=self.pipeline.documents
         )
-        self.pipeline.df_test, self.pipeline.test_label_feature_list = self.pipeline.feature_function(
-            documents=self.pipeline.test_documents
-        )
```
Comment (da-konfuzio) [inline tests/trainer/test_information_extraction.py:99]: How does the pipeline do evaluation without features to the test docs?

File: tests/trainer/test_information_extraction.py [hunk 6]
```diff
@@ -138,142 +145,489 @@ def test_1_configure_pipeline(self):
         self.pipeline.documents = self.pipeline.category.documents()[:5]
         self.pipeline.test_documents = self.pipeline.category.test_documents()[:1]
         # todo have a separate test case for calculating features of offline documents
-        for doc in self.pipeline.documents + self.pipeline.test_documents:
-            doc.set_offline()
+        # for doc in self.pipeline.documents + self.pipeline.test_documents:
+        #     doc.set_offline()
```
Comment (da-konfuzio) [inline tests/trainer/test_information_extraction.py:149]: why is this necessary?

File: konfuzio_sdk/trainer/information_extraction.py [hunk 7]
```diff
@@ -2612,7 +2705,25 @@ def lose_weight(self):
...
+            if span.end_offset < doc_spans[s_i].start_offset:
+                continue
+
+            r = range(doc_spans[s_i].start_offset, doc_spans[s_i].end_offset + 1)
+            if span.start_offset in r and span.end_offset in r:
+                span.annotation.label = doc_spans[s_i].annotation.label
+
+    def feature_function(
+        self, documents: List[Document], no_label_limit=None, retokenize=True
+    ) -> Tuple[List[pandas.DataFrame], list]:
```
Comment (da-konfuzio) [inline konfuzio_sdk/trainer/information_extraction.py:2726]: Can you add a short documentation in the docstring for each param? Especially for the `retokenize` option given how tricky it was to find out that this was necessary :slightly_smiling_face:

**Comment by da-konfuzio** [comment]
In the notebook:
- "This is the output of Object Character Recognition (OCR) model on the following image:" (I think you meant Optical Character Recognition :wink: )
- Towards the end, where we display the output of the model, I would merge a few cells or add a bit more comments
- Overall looks good to me

**Comment by da-konfuzio** [comment]
About the failing testcase ([test_size_of_project](https://github.com/konfuzio-ai/konfuzio-sdk/blob/934dd1c5318d3d172bd0170c8d7a424639ed44a2/tests/test_data.py#L1399)), the new value can just be 1.74 since project 46 on app was changed when we created new docs in order to compare performance between Trainer and SDK.

However, I'm not sure I understand the point of this testcase. @atraining do you know more?

File: tests/trainer/test_information_extraction.py (hunk 5 above)
Comment (samuel500) [reply tests/trainer/test_information_extraction.py:99]: Features are created in the `extract` method:
https://github.com/konfuzio-ai/konfuzio-sdk/blob/9435-maximum-strict-evaluation-score/konfuzio_sdk/trainer/information_extraction.py#L2598

File: tests/trainer/test_information_extraction.py [hunk 8]
```diff
@@ -346,19 +767,292 @@ def test_extract_with_correctly_fitted_clf(self):
...
+        assert len(annotation.spans) == 1
+
+    def test_4_span_attributes_of_annotation_created(self):
+        """Test attributes of the span in the annotation created in the sample document."""
+        annotation = self.sample_document.annotations(use_correct=False)[0]
+        assert annotation.spans[0].start_offset == self.extraction_df.loc[0, 'Start']
+        assert annotation.spans[0].end_offset == self.extraction_df.loc[0, 'End']
+        # The document used does not have bounding boxes, so we cannot have the coordinates
+        assert annotation.spans[0].offset_string == 'pizza'
+        assert annotation.spans[0].x0 is None
```
Comment (atraining) [inline tests/trainer/test_information_extraction.py:870]: why is it None when it should be 10 ?

File: tests/trainer/test_information_extraction.py [hunk 9]
```diff
@@ -346,19 +767,292 @@ def test_extract_with_correctly_fitted_clf(self):
...
+            label=self.label,
+            label_set=self.label_set,
+            annotation_set=annotation_set_1,
+        )
+
+        annotation = document.annotations(use_correct=False)[0]
+        # The document used is an empty document, therefore it does not have text or bounding boxes,
+        # so we cannot have the offset string or the coordinates
+        assert annotation.spans[0].offset_string is None
+        assert annotation.spans[0].x0 is None
```
Comment (atraining) [inline tests/trainer/test_information_extraction.py:919]: do you want to allow this behavior or raise a NotImplementedError for now?

File: tests/trainer/test_information_extraction.py [hunk 10]
```diff
@@ -346,19 +767,292 @@ def test_extract_with_correctly_fitted_clf(self):
...
+
+        with self.assertRaises(ValueError) as context:
+            add_extractions_as_annotations(
+                extractions=extraction_df,
+                document=document,
+                label=self.label,
+                label_set=self.label_set,
+                annotation_set=annotation_set_1,
+            )
+            assert 'Extraction do not contain all required fields' in context.exception
```
Comment (atraining) [inline tests/trainer/test_information_extraction.py:940]: use match or move out of the context manager https://stackoverflow.com/a/56569533/5800596

File: tests/test_data.py [hunk 11]
```diff
@@ -1396,7 +1396,7 @@ def _getsize(obj):
         for document in prj.documents:
             document.text
         after = _getsize(prj)
-        assert after / before > 1.8
+        assert 1.7 < after / before < 1.8
```
[the remark this answers is not in the document]
Comment (samuel500) [reply tests/test_data.py:1399]: Do you mean checking for the absolute number `after-before` rather than the ratio?

[the remark this answers is not in the document]
Comment (da-konfuzio) [reply tests/test_data.py:1399]: @atraining I don't get your comment

File: tests/trainer/test_information_extraction.py [hunk 12]
```diff
@@ -67,63 +74,94 @@ def display_top(snapshot, key_type='lineno', limit=30):
...
         """Set up the Data and Pipeline."""
-        tracemalloc.start()
         cls.project = Project(id_=None, project_folder=OFFLINE_PROJECT)
-        cls.pipeline = DocumentAnnotationMultiClassModel()
-        display_top(tracemalloc.take_snapshot())
-
-    def tearDown(self) -> None:
-        """Print a Snapshot after running a test."""
-        display_top(tracemalloc.take_snapshot())
+        cls.pipeline = DocumentEntityMulticlassModel()
```
[the remark this answers is not in the document]
Comment (samuel500) [reply tests/trainer/test_information_extraction.py:109]: Will be changed after: https://gitlab.com/konfuzio/objectives/-/issues/9511

File: tests/test_evaluate.py [hunk 13]
```diff
@@ -902,15 +902,15 @@ def test_true_positive(self):
...
+        assert evaluation.tp() == 5
 
     def test_false_positive(self):
         """Count zero Annotations from two Training Documents."""
         project = LocalTextProject()
         true_document = project.documents[0]
         predicted_document = project.test_documents[0]
         evaluation = Evaluation(documents=list(zip([true_document], [predicted_document])))
-        assert evaluation.fp() == 2
+        assert evaluation.fp() == 3
```
[the remark this answers is not in the document]
Comment (da-konfuzio) [reply tests/test_evaluate.py:913]: Probably there is no way other than serializing the evaluation. Adding custom logic in the testcase in order to retrieve the expected number would basically replicate the FP calculation already present within the EvaluationCalculator.

File: tests/test_data.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (atraining) [reply tests/test_data.py:1399]: just a boilerplate to calculate the lower limit

`_getsize(''.join([doc.text for doc in prj.documents])) + before < after < `

[the remark this answers is not in the document]
Comment (samuel500) [reply tests/test_data.py:1399]: `assert _getsize([doc.text for doc in prj.documents]) + before < after + 500` I what I got to work here. There seems to be some small memory efficiency gains when the string are stored in the document, or some other data gets deleted (?), which makes `after` slightly smaller than `[doc.text for doc in prj.documents]) + before`. Joining the strings makes it worse.

File: konfuzio_sdk/trainer/information_extraction.py [hunk 14]
```diff
@@ -1718,10 +1718,92 @@ def generate_feature_dict_from_occurence_dict(occurence_dict, catchphrase_list,
...
+                document=document,
+                label=label,
+                confidence=annotation['Accuracy'],
+                label_set=label_set,
+                annotation_set=annotation_set,
+                spans=[Span(start_offset=annotation['Start'], end_offset=annotation['End'])],
+            )
+
+
+def extraction_result_to_document(document: Document, extraction_result: dict) -> Document:
```
[the remark this answers is not in the document]
Comment (da-konfuzio) [reply konfuzio_sdk/trainer/information_extraction.py:1711]: Extraction result dictionaries have LabelSet names (or Label names) in the keys, and then Dataframes in the values, so an unexpected dictionary would be one that contains non existing LabelSet/Label names in the keys, which wouldn't be retrievable by the used methods such as get_label_set_by_name (raises IndexError). Or there could be an invalid Dataframe or an object that is not a Dataframe, which would currently crash.

I added testcases to show that we already cover the LabelSet/Label checks, and added checks for the Dataframe format and type.

File: tests/tokenizers/test_base.py [hunk 15]
```diff
@@ -172,6 +172,24 @@ def test_evaluate_output_offsets_with_document(self):
...
+    def test_evaluate_dataset_output_without_test_documents(self):
+        """Test evaluate an empty list of Documents."""
+        with self.assertRaises(ValueError):
+            self.tokenizer.evaluate_dataset([])
+
+    def test_evaluate_dataset_output_with_test_documents(self):
+        """Test evaluate a Category with a test Documents."""
+        result = self.tokenizer.evaluate_dataset(self.category_2.test_documents())
+        assert len(result.data) == 2
+        assert result.data.category_id.dropna().unique() == self.category_2.id_
```
[the remark this answers is not in the document]
Comment (da-konfuzio) [reply tests/tokenizers/test_base.py]: Done, it was because an empty span for the NO_LABEL_SET is always created

File: konfuzio_sdk/trainer/information_extraction.py [hunk 16]
```diff
@@ -2722,8 +2722,8 @@ def feature_function(
                     ):
                         pass
                     else:
-                        logger.error(
-                            f"{span.annotation} is unrevised in this dataset and may impact model performance!"
+                        raise ValueError(
+                            f"{span.annotation} is unrevised in this dataset and can't be used for training!"
```
Comment (atraining) [inline konfuzio_sdk/trainer/information_extraction.py:2726]: @da-konfuzio  To clarify: I am fine to use log errors to keep the Extraction AI failing rate low. otherwise, the training process should fail and tell the user via the Server Interface, what needs to be done to fix this: "Please revise the following Annotations: link1, link2 ..."

Comment (da-konfuzio) [reply konfuzio_sdk/trainer/information_extraction.py]: Ok then I'll keep it as an option with `require_revised_annotations=False` as default which would print error logs, otherwise it raises a ValueError.
